File: common/src/main/java/com/nghiatl/common/service/LocationService.java
package com.nghiatl.common.service;

import android.Manifest;
import android.app.Activity;
import android.app.Service;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.location.Location;
import android.location.LocationListener;
import android.location.LocationManager;
import android.os.Bundle;
import android.os.IBinder;
import android.util.Log;

import androidx.core.app.ActivityCompat;

/**
 * Created by Nghia-PC on 7/16/2015.
 * Lấy GPS Thông qua Service - Không có tùy
 * Example:
 * LocationService location = new LocationService(mContext);
 *         if (!location.canGetLocation()) {
 *             // Không lấy được vị trí
 *
 *         } else {
 *             // lấy được vị trí
 *             location.getLatitude()
 *         }
 */
public class LocationService extends Service {

    private static final int MY_PERMISSION_ACCESS_COURSE_LOCATION = 99;
    Location mLocation;  // Dữ liệu trả về
    boolean mIsGPSEnabled = false;  // flag for GPS status
    boolean mIsNetworkEnabled = false;  // flag for network status
    boolean mCanGetLocation = false;

    private static final long DEFAULT_MIN_METER_UPDATE_DISTANCE = 10; // 10 meters
    private static final long DEFAULT_MIN_MILLISECOND_UPDATE_TIME = 1; // 1/1000s
    /** Khoảng cách Update */
    private long minMeterUpdateDistance;
    /** Thời gian Update */
    private long minMillisecondUpdateTime;
    private final Context mContext;
    protected LocationManager mLocationManager;
    LocationListener mLocationListener = new LocationListener() {
        @Override
        public void onLocationChanged(Location location) {
            mLocation = location;
        }

        @Override
        public void onStatusChanged(String provider, int status, Bundle extras) {

        }

        @Override
        public void onProviderEnabled(String provider) {

        }

        @Override
        public void onProviderDisabled(String provider) {

        }
    };


    public LocationService(Context context) {
        this.mContext = context;
        mLocation = new Location("Some Where");
        mLocation.setLatitude(0.0);
        mLocation.setLongitude(0.0);

        minMeterUpdateDistance = DEFAULT_MIN_METER_UPDATE_DISTANCE;
        minMillisecondUpdateTime = DEFAULT_MIN_MILLISECOND_UPDATE_TIME;
        getLocation();
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        return super.onStartCommand(intent, flags, startId);
    }

    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }

    @Override
    public boolean onUnbind(Intent intent) {
        return super.onUnbind(intent);
    }


    @Override
    public void onDestroy() {
        super.onDestroy();


        mLocationListener = null;
    }

    public Location getLocation() {
        try {
            mLocationManager = (LocationManager) mContext.getSystemService(LOCATION_SERVICE);

            mIsGPSEnabled = mLocationManager.isProviderEnabled(LocationManager.GPS_PROVIDER);
            mIsNetworkEnabled = mLocationManager.isProviderEnabled(LocationManager.NETWORK_PROVIDER);

            if (!mIsGPSEnabled && !mIsNetworkEnabled) {
                // Không lấy được GPS
            } else {
                this.mCanGetLocation = true;
                if (mIsNetworkEnabled) {
                    getLocationByNetwork();
                }
                if (mIsGPSEnabled) {
                    getLocationByGPS();
                }
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
        return mLocation;
    }

    public double getLatitude() {
        if (mLocation != null) {
            return mLocation.getLatitude();
        } else {
            return 0;
        }
    }

    public double getLongitude() {
        if (mLocation != null) {
            return mLocation.getLongitude();
        } else {
            return 0;
        }
    }

    public boolean canGetLocation() {
        return this.mCanGetLocation;
    }

    /** Dừng GPS */
    public void stopUsingGPS() {
        //Check + Request Permission
        if (ActivityCompat.checkSelfPermission(mContext, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED
                && ActivityCompat.checkSelfPermission(mContext, Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions((Activity)mContext, new String[] {  android.Manifest.permission.ACCESS_COARSE_LOCATION  },
                    MY_PERMISSION_ACCESS_COURSE_LOCATION);
            return;
        }

        if (mLocationManager != null) {
            mLocationManager.removeUpdates(mLocationListener);
        }
    }


    /** Lấy vị trí qua GPS */
    private void getLocationByGPS() {
        Log.d("Location", "Get by GPS");

        //Check + Request Permission
        if (ActivityCompat.checkSelfPermission(mContext, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED
                && ActivityCompat.checkSelfPermission(mContext, Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions((Activity)mContext, new String[] {  android.Manifest.permission.ACCESS_COARSE_LOCATION  },
                    MY_PERMISSION_ACCESS_COURSE_LOCATION);
            return;
        }

        if (mLocation == null) {
            mLocationManager.requestLocationUpdates(
                    LocationManager.GPS_PROVIDER,
                    minMillisecondUpdateTime,
                    minMeterUpdateDistance,
                    mLocationListener);
            if (mLocationManager != null) {
                Location location = mLocationManager.getLastKnownLocation(LocationManager.GPS_PROVIDER);
                if (location != null) mLocation = location;
            }
        }
    }

    /** Lấy vị trí Qua mạng */
    private void getLocationByNetwork() {
        Log.d("Location", "Get by Network");

        //Check + Request Permission
        if (ActivityCompat.checkSelfPermission(mContext, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED
                && ActivityCompat.checkSelfPermission(mContext, Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions((Activity)mContext, new String[] {  android.Manifest.permission.ACCESS_COARSE_LOCATION  },
                    MY_PERMISSION_ACCESS_COURSE_LOCATION);
            return;
        }

        mLocationManager.requestLocationUpdates(
                LocationManager.NETWORK_PROVIDER,
                minMillisecondUpdateTime,
                minMeterUpdateDistance,
                mLocationListener);

        if (mLocationManager != null) {
            Location location = mLocationManager.getLastKnownLocation(LocationManager.NETWORK_PROVIDER);
            if (location != null) mLocation = location;
        }
    }
}

File: common/src/main/java/com/nghiatl/common/service/LocationService.kt
package com.nghiatl.common.service

import android.Manifest
import android.annotation.SuppressLint
import android.app.Activity
import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.content.pm.PackageManager
import android.location.Location
import android.location.LocationListener
import android.location.LocationManager
import android.net.Uri
import android.os.Bundle
import android.os.Looper
import android.provider.Settings
import android.util.Log
import androidx.core.app.ActivityCompat

private const val MY_PERMISSION_ACCESS_COURSE_LOCATION = 99
private const val DEFAULT_MIN_METER_UPDATE_DISTANCE: Long = 10 // 10 meters
private const val DEFAULT_MIN_MILLISECOND_UPDATE_TIME: Long = 1 // 1/1000s

class LocationService(
    private val context: androidx.fragment.app.FragmentActivity,
    private val updateRealtime: Boolean = false,
    var listener: Listener? = null
) {
    private val LASTEST_LOCATION_NAME: String = "LASTEST_LOCATION_NAME"
    private val KEY_LATITUDE: String = "KEY_LATITUDE"
    private val KEY_LONGITUDE: String = "KEY_LONGITUDE"
    private val TAG: String by lazy { this.javaClass.simpleName }

    var mlocation: Location? = null
    var canGetLocation = false
    var minMeterUpdateDistance: Long = DEFAULT_MIN_METER_UPDATE_DISTANCE
    var minMillisecondUpdateTime: Long = DEFAULT_MIN_MILLISECOND_UPDATE_TIME

    // V1
    private var locationManager: LocationManager? = null
    private val locationManagerListener: LocationListener = object : LocationListener {
        override fun onLocationChanged(location: Location) {
            updateLocation(location, true)
        }

        override fun onStatusChanged(provider: String, status: Int, extras: Bundle) {}
        override fun onProviderEnabled(provider: String) {}
        override fun onProviderDisabled(provider: String) {}
    }

    // V2
    /*private var locationProviderClient: FusedLocationProviderClient? = null
    private val locationProviderListener = object : LocationCallback() {
        override fun onLocationResult(locationResult: LocationResult?) {
            locationResult?.let { locationResult ->
                updateLocation(locationResult.lastLocation, true)
            }
        }
    }
    private var locationRequest: LocationRequest? = null*/


    init {
        getLocationBySystemService()    // todo remove
        /*Dexter.withActivity(context)
            .withPermissions(
                Manifest.permission.ACCESS_FINE_LOCATION,
                Manifest.permission.ACCESS_COARSE_LOCATION
            ).withListener(object : MultiplePermissionsListener {
                override fun onPermissionsChecked(report: MultiplePermissionsReport?) {
                    report?.let {
                        if (report.areAllPermissionsGranted()) {
                            getLocationBySystemService()
                            Timber.e("$TAG : onPermissionsChecked allow")
                        } else {
                            Timber.e("$TAG : onPermissionsChecked not allow")
                            showRequestPermissionDialog()
                            updateLocation(location, false)
                        }
                    }
                }

                override fun onPermissionRationaleShouldBeShown(permissions: MutableList<PermissionRequest>?, token: PermissionToken?) {
                    token?.continuePermissionRequest()  //request location again
                }
            })
            .withErrorListener {
                Timber.e("$TAG : withErrorListener")
            }.check()*/
    }

    /*fun showRequestPermissionDialog() {
        AppDialog.showMessageDialog(
            context.supportFragmentManager, false,
            context.getString(R.string.permission_request_location_title),
            context.getString(R.string.permission_request_location_message),
            context.getString(R.string.permission_settings),
            { view, dialog ->
                // goto setting
                openSettings()
                dialog.dismiss()
            },
            context.getString(R.string.common_cancel),
            { view, dialog -> dialog.dismiss() }
        )
    }*/

    // navigating user to app settings
    /*private fun openSettings() {
        val intent = Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS)
        val uri: Uri = Uri.fromParts("package", context.packageName, null)
        intent.data = uri
        context.startActivityForResult(intent, 101)
    }*/

    @SuppressLint("MissingPermission")
    private fun getLocationBySystemService() {
        var location: Location? = null
        try {
            locationManager = context.getSystemService(Context.LOCATION_SERVICE) as LocationManager
            locationManager?.let { locationManager ->
                if (validateLocationPermission()) {
                    //val gpsProvider = if (isNetworkEnabled) LocationManager.NETWORK_PROVIDER else LocationManager.GPS_PROVIDER

                    // check all providers can get GPS
                    for (gpsProvider in locationManager.allProviders) {
                        location = locationManager.getLastKnownLocation(gpsProvider)
                        if (location != null) {
                            updateLocation(location!!, true)

                            // Auto update when gps change
                            if (updateRealtime) {
                                locationManager.requestLocationUpdates(
                                    gpsProvider,
                                    minMillisecondUpdateTime,
                                    minMeterUpdateDistance.toFloat(),
                                    locationManagerListener)
                            }
                            break
                        }
                    }
                }
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }

        // call get gps by method 2
        if (this.mlocation == null) {
            //getLocationByLocationService()
            updateLocation(canGetLocation = false)  // todo remove
        }
    }

    /*@SuppressLint("MissingPermission")
    fun getLocationByLocationService() {
        if (!validateLocationPermission()) return

        locationProviderClient = LocationServices.getFusedLocationProviderClient(context)
        locationProviderClient?.lastLocation?.addOnSuccessListener { location: Location? ->
            if (location != null) {
                updateLocation(location, true)

                // Auto update when gps change
                if (updateRealtime) {
                    locationRequest = LocationRequest.create()
                    locationRequest?.priority = LocationRequest.PRIORITY_HIGH_ACCURACY
                    locationRequest?.interval = minMillisecondUpdateTime
                    locationProviderClient?.requestLocationUpdates(LocationRequest(), locationProviderListener, Looper.myLooper())
                }
            } else {
                updateLocation(canGetLocation = false)
            }
        }?.addOnFailureListener {
            *//*when ((it as ApiException).statusCode) {
                LocationSettingsStatusCodes.RESOLUTION_REQUIRED -> {

                }
                LocationSettingsStatusCodes.SETTINGS_CHANGE_UNAVAILABLE -> {

                }
            }*//*
            updateLocation(canGetLocation = false)
        }
    }*/

    private fun updateLocation(location: Location = Location(TAG), canGetLocation: Boolean) {
        this.canGetLocation = canGetLocation
        this.mlocation = location
        saveLatestLocation(location)
        listener?.onLocationChanged(location)
        Log.d(TAG, "latitude: ${location.latitude}, longitude: ${location.longitude}")
    }

    val location: Location
        get() {
            return if (mlocation != null) {
                mlocation!!
            } else {
                Location(TAG).apply {
                    latitude = getLatestLatitude()
                    longitude = getLatestLongitude()
                }
            }
        }

    val latitude: Double
        get() = this.mlocation?.latitude ?: getLatestLatitude()

    val longitude: Double
        get() = this.mlocation?.longitude ?: getLatestLongitude()

    @SuppressLint("MissingPermission")
    fun stopUsingGPS() {
        if (!validateLocationPermission()) return

        locationManager?.removeUpdates(locationManagerListener)
        //locationProviderClient?.removeLocationUpdates(locationProviderListener)
    }

    private fun validateLocationPermission(): Boolean {
        //Check + Request Permission
        if (ActivityCompat.checkSelfPermission(context, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED
            && ActivityCompat.checkSelfPermission(context, Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions((context as Activity), arrayOf(Manifest.permission.ACCESS_COARSE_LOCATION),
                MY_PERMISSION_ACCESS_COURSE_LOCATION)
            return false
        }
        return true
    }

    private fun saveLatestLocation(location: Location) {
        context.getSharedPreferences(LASTEST_LOCATION_NAME, Context.MODE_PRIVATE).edit()
            .putString(KEY_LATITUDE, location.latitude.toString())
            .putString(KEY_LONGITUDE, location.longitude.toString())
            .apply()
    }

    private fun getLatestLatitude() : Double {
        val latitudeStr = context.getSharedPreferences(LASTEST_LOCATION_NAME, Context.MODE_PRIVATE)
            .getString(KEY_LATITUDE, "0")
        return latitudeStr?.toDouble() ?: 0.0
    }

    private fun getLatestLongitude() : Double {
        val latitudeStr = context.getSharedPreferences(LASTEST_LOCATION_NAME, Context.MODE_PRIVATE)
            .getString(KEY_LONGITUDE, "0")
        return latitudeStr?.toDouble() ?: 0.0
    }

    interface Listener {
        fun onLocationChanged(location: Location)
    }
}